File: playwright-typescript/pages/payment-authentication-page.ts
import { Locator, Page } from '@playwright/test';

export class PaymentAuthenticationPage {
  readonly page: Page;
  readonly checkoutFrame;
  readonly authenticationFrame;
  readonly completeButton: Locator;
  readonly failButton: Locator;

  constructor(page: Page) {
    this.page = page;
    this.checkoutFrame = page.frameLocator('#checkout-demo');
    this.authenticationFrame = this.checkoutFrame.frameLocator('iframe[src^="https://js.stripe.com/v3/three-ds-2-challenge"]').nth(0).frameLocator('iframe#challengeFrame');
    this.completeButton = this.authenticationFrame.locator('#test-source-authorize-3ds');
    this.failButton = this.authenticationFrame.locator('#test-source-fail-3ds');
  }

  async complete() {
    await this.completeButton.click();
  }

  async fail() {
    await this.failButton.click();
  }
}

File: selenium-ruby/features/pages/payment_authentication_page.rb
require_relative 'base_page'

class PaymentAuthenticationPage < BasePage
  def id
    'challengeFrame'
  end

  def complete
    $driver.execute_script('arguments[0].click();', complete_button)
  end

  def fail
    $driver.execute_script('arguments[0].click();', fail_button)
  end

  private

  def complete_button
    find_ele('test-source-authorize-3ds')
  end

  def fail_button
    find_ele('test-source-fail-3ds')
  end
end
